### ISSUE TITLE
RPS issues in console output and likely report for inject per second tests

### ISSUE THREAD

**Issue by seertenedos** [PragmaticFlow/NBomber#262]
If i understand the inject per second tests and set a rate of say 100 it means it should kick off 100 requests per second. I would also assume the RPS = request count all / duration in seconds but i have noticed that in higher RPS situations this is not true.

If I use constant load set with 100 threads i get an RPS of ~600 and about ~36.5k request count on GCP CloudRun.

If i do exact same test but set the requests per second to 600 instead of fixed nbr threads things get interesting.  In that case i get an RPS of just 437 but the request count is still around 35.8 which to my calculation means the RPS should have shown ~578. I captured the output of a 60 second test for this below.

The more i scale the service the worse it got. I made the backend support around 1.5k using 100 threads but when i used an RPS of 1.2k the result was around the same request count but the RPS was extremely low in the low 200-300 RPS which made no sense to me.

Test using 100 threads
```
100.00% simulation: 'keep_constant', copies: '100'                                                              00:00:11
100.00% simulation: 'keep_constant', copies: '100'                                                              00:01:01
100.00% closing connections for connection pool: 'Lorenzo-CreateSridOnly.LorenzoClient'                         00:00:00

scenario: 'Lorenzo-CreateSridOnly', duration: '00:01:00', ok count: '36626', fail count: '0', all data: '0.0' MB

+----------------------+--------------------------------------------------------------+
| step                 | details                                                      |
+----------------------+--------------------------------------------------------------+
| - name               | CreateSrid1                                                  |
+----------------------+--------------------------------------------------------------+
| - request count      | all = 36626 | OK = 36626 | failed = 0                        |
+----------------------+--------------------------------------------------------------+
| - latency            | RPS = 607 | min = 11 | mean = 151 | max = 4574               |
+----------------------+--------------------------------------------------------------+
| - latency percentile | 50% = 106 | 75% = 188 | 95% = 372 | 99% = 784 | StdDev = 207 |
+----------------------+--------------------------------------------------------------+
```

Test using 600 RPS
```
100.00% simulation: 'inject_per_sec', rate: '600'                                                               00:00:11
100.00% simulation: 'inject_per_sec', rate: '600'                                                               00:01:02
100.00% closing connections for connection pool: 'Lorenzo-CreateSridOnly.LorenzoClient'                         00:00:00
scenario: 'Lorenzo-CreateSridOnly', duration: '00:01:00', ok count: '35861', fail count: '0', all data: '0.0' MB
+----------------------+------------------------------------------------------------------+
| step                 | details                                                          |
+----------------------+------------------------------------------------------------------+
| - name               | CreateSrid1                                                      |
+----------------------+------------------------------------------------------------------+
| - request count      | all = 35861 | OK = 35861 | failed = 0                            |
+----------------------+------------------------------------------------------------------+
| - latency            | RPS = 437 | min = 14 | mean = 1094 | max = 17718                 |
+----------------------+------------------------------------------------------------------+
| - latency percentile | 50% = 505 | 75% = 1149 | 95% = 3929 | 99% = 7203 | StdDev = 1618 |
+----------------------+------------------------------------------------------------------+
```

**Comment by AntyaDev** [comment]
Hi @seertenedos 
Why on your first test you have used simulation: 'keep_constant' and then on the second test you have used simulation: 'inject_per_sec'?

**Comment by seertenedos** [comment]
because i wanted to know the limit of my system and the inject_per_sec for my webservice was producing numbers that did not match up with what google cloud said the service was receiving so i tested both. If using multi threads i can get one RPS but the service is only at 80% then surely i should be able to get similar or better RPS from inject_per_sec or am i misunderstanding it.

**Comment by AntyaDev** [comment]
can you please place here results for inject-per sec only with 100 req per sec and 600 just to compare

**Comment by seertenedos** [comment]
100 RPS
```
test suite: 'nbomber_default_test_suite_name'
test name: 'nbomber_default_test_name'
scenario: 'Lorenzo-CreateSridOnly', duration: '00:01:00', ok count: '6000', fail count: '0', all data: '0.0' MB
+----------------------+--------------------------------------------------------------+
| step                 | details                                                      |
+----------------------+--------------------------------------------------------------+
| - name               | CreateSrid1                                                  |
+----------------------+--------------------------------------------------------------+
| - request count      | all = 6000 | OK = 6000 | failed = 0                          |
+----------------------+--------------------------------------------------------------+
| - latency            | RPS = 98 | min = 18 | mean = 154 | max = 6214                |
+----------------------+--------------------------------------------------------------+
| - latency percentile | 50% = 82 | 75% = 128 | 95% = 552 | 99% = 1188 | StdDev = 283 |
+----------------------+--------------------------------------------------------------+
```
600 RPS - Had connection issue to DB so errors but should have been RPS above 98 based on test above. It seems that errors someone greatly reduce the RPS for successful requests being reported. if it was 13 RPS then OK count should have only been 780 not 15421. 
```
test suite: 'nbomber_default_test_suite_name'
test name: 'nbomber_default_test_name'
scenario: 'Lorenzo-CreateSridOnly', duration: '00:01:00', ok count: '15421', fail count: '13952', all data: '0.0' MB
+----------------------+-----------------------------------------------------------------------+
| step                 | details                                                               |
+----------------------+-----------------------------------------------------------------------+
| - name               | CreateSrid1                                                           |
+----------------------+-----------------------------------------------------------------------+
| - request count      | all = 29373 | OK = 15421 | failed = 13952                             |
+----------------------+-----------------------------------------------------------------------+
| - latency            | RPS = 13 | min = 127 | mean = 10178 | max = 35901                     |
+----------------------+-----------------------------------------------------------------------+
| - latency percentile | 50% = 11071 | 75% = 12223 | 95% = 16799 | 99% = 22751 | StdDev = 4776 |
+----------------------+-----------------------------------------------------------------------+
```
600 RPS - slightly scaling but basically running normally 
```
test suite: 'nbomber_default_test_suite_name'
test name: 'nbomber_default_test_name'
scenario: 'Lorenzo-CreateSridOnly', duration: '00:01:00', ok count: '35852', fail count: '0', all data: '0.0' MB
+----------------------+-----------------------------------------------------------------+
| step                 | details                                                         |
+----------------------+-----------------------------------------------------------------+
| - name               | CreateSrid1                                                     |
+----------------------+-----------------------------------------------------------------+
| - request count      | all = 35852 | OK = 35852 | failed = 0                           |
+----------------------+-----------------------------------------------------------------+
| - latency            | RPS = 497 | min = 20 | mean = 744 | max = 16367                 |
+----------------------+-----------------------------------------------------------------+
| - latency percentile | 50% = 340 | 75% = 618 | 95% = 2773 | 99% = 7591 | StdDev = 1421 |
+----------------------+-----------------------------------------------------------------+
```

**Comment by AntyaDev** [comment]
Hi @seertenedos when you have fails, NBomber will not take them for statistics, that's why you have so small RPS.
When fail happens it's not reasonable to take its latency for statistics since the error could give you 1 ms which completely ruins your overall stats like min, max. In the next release, I will print some hints in the console when we got at least one error. Your load tests should have no errors.

**Comment by seertenedos** [comment]
Hi @AntyaDev i have to disagree a little. A failed request should be ignored and should not count as a negative value for calculating RPS. Yes it would be great if errors never happen but most systems will start to have issues when you push them past their limits which is one of the use cases for a load test (to find those limits). 

If take this result below
```
test suite: 'nbomber_default_test_suite_name'
test name: 'nbomber_default_test_name'
scenario: 'Lorenzo-CreateSridOnly', duration: '00:01:00', ok count: '15421', fail count: '13952', all data: '0.0' MB
+----------------------+-----------------------------------------------------------------------+
| step                 | details                                                               |
+----------------------+-----------------------------------------------------------------------+
| - name               | CreateSrid1                                                           |
+----------------------+-----------------------------------------------------------------------+
| - request count      | all = 29373 | OK = 15421 | failed = 13952                             |
+----------------------+-----------------------------------------------------------------------+
| - latency            | RPS = 13 | min = 127 | mean = 10178 | max = 35901                     |
+----------------------+-----------------------------------------------------------------------+
| - latency percentile | 50% = 11071 | 75% = 12223 | 95% = 16799 | 99% = 22751 | StdDev = 4776 |
+----------------------+-----------------------------------------------------------------------+
```
It ran for 60 seconds and had 15,421 successful result so successful RPS of 257 and 13,952 failures so a failure RPS of 232. Even if you are taking one away from the other the numbers of 13 for the RPS still don't add up and i would prefer a separate row for RPS if you are not just doing an RPS of successful requests. I think the other RPS values you are maybe using could be useful to see but i found the overall RPS of 13 when in fact success RPS was 257 was a little confusing. There really is no explanation of where the 13 RPS comes from.

**Comment by tienle488** [comment]
Hello @AntyaDev 

Not sure that I should initiate a new issue but I have a similar concern.  My test runs well and from reviewing the receiving end of my tests, the number of messages posted to my Rabbit MQ for example looks correct; however, the "all, OK, and failed +RPS" counts don't quite match up with count reported by Rabbit MQ.  Reference the attachment:
![NBomberReportingRequestsMadeNotAccurate_RabbitMqLooksCorrect](https://user-images.githubusercontent.com/23621965/108783466-65ab6000-7522-11eb-90ff-1f6d76aa5d47.PNG)
  
Note: I'm confident that the Rabbit MQ is truthful because when I run a similar test using Visual Studio Load Test load testing my .NET Framework 4.7.2 telemetry client, I get 25K metrics as expected.  At the same time, VSLT does not report grand total so we really haven't relied on VSLT for this actually.  Hum...
![VSLTDoesn'tReportGrandTotalButRabbitMqLooksCorrect](https://user-images.githubusercontent.com/23621965/108785545-7c53b600-7526-11eb-9196-76626958dc22.PNG)

Thanks,
Tien

**Comment by AntyaDev** [comment]
Hi @seertenedos

Recently we released a new version of NBomber 2.0.0 rc5 that fixes many performance bugs.
Can you please try it out on your test?

**Comment by seertenedos** [comment]
Sorry i have moved on to other things now so i can't actually test this any
more sadly.
